File: scripts/import_payment_bulk.py
#!/usr/bin/env python3
"""
Fast bulk payment CSV import using SQLAlchemy + MySQL upsert.

Usage:
  pip install -r requirements.txt
  python scripts/import_payment_bulk.py jodo "path/to/file.csv"

Supports: jodo, razorpay, early, offline, bank, propelld, others
"""
from __future__ import annotations

import argparse
import os
import re
import sys
from datetime import datetime
from pathlib import Path
from urllib.parse import quote_plus, unquote, urlparse

import pandas as pd
from dotenv import load_dotenv
from sqlalchemy import create_engine, text
from sqlalchemy.engine import Engine

ROOT = Path(__file__).resolve().parents[1]
load_dotenv(ROOT / ".env")

PAYMENT_TABLES = {
    "razorpay": ("RazorpayPayment", "Razorpay"),
    "jodo": ("JodoPayment", "Jodo"),
    "early": ("EarlyPayment", "Early"),
    "offline": ("OfflinePayment", "Offline"),
    "bank": ("BankPayment", "Bank"),
    "propelld": ("PropelldPayment", "Propelld"),
    "others": ("OthersPayment", "Others"),
}

CHUNK_SIZE = 2000


def prisma_url_to_sqlalchemy(url: str) -> str:
    """mysql://user:pass@host:port/db -> mysql+pymysql://..."""
    if url.startswith("mysql+pymysql://"):
        return url
    if not url.startswith("mysql://"):
        raise ValueError("DATABASE_URL must be a mysql:// connection string")
    return "mysql+pymysql://" + url[len("mysql://") :]


def parse_date(value) -> datetime | None:
    if value is None or (isinstance(value, float) and pd.isna(value)):
        return None
    s = str(value).strip()
    if not s:
        return None
    for fmt in ("%Y-%m-%d", "%d-%m-%Y", "%d/%m/%Y", "%m/%d/%Y"):
        try:
            return datetime.strptime(s.split()[0], fmt)
        except ValueError:
            continue
    return None


def parse_float(value) -> float:
    if value is None or (isinstance(value, float) and pd.isna(value)):
        return 0.0
    s = str(value).strip().replace(",", "")
    if not s:
        return 0.0
    try:
        return float(s)
    except ValueError:
        return 0.0


def parse_int_optional(value) -> int | None:
    if value is None or (isinstance(value, float) and pd.isna(value)):
        return None
    s = str(value).strip()
    if not s:
        return None
    try:
        return int(float(s))
    except ValueError:
        return None


def first_col(row: pd.Series, *names: str):
    for name in names:
        if name in row.index:
            val = row[name]
            if pd.notna(val) and str(val).strip() != "":
                return val
    return None


def load_reference_maps(engine: Engine) -> tuple[set[int], dict[int, int | None]]:
    with engine.connect() as conn:
        enroll_rows = conn.execute(text("SELECT id FROM Enrollment")).fetchall()
        valid_ids = {int(r[0]) for r in enroll_rows}

        batch_rows = conn.execute(
            text(
                "SELECT enrollmentId, batchId FROM AdmissionForm WHERE enrollmentId IS NOT NULL"
            )
        ).fetchall()
        batch_map = {int(r[0]): (int(r[1]) if r[1] is not None else None) for r in batch_rows}
    return valid_ids, batch_map


def normalize_csv(df: pd.DataFrame) -> pd.DataFrame:
    df = df.copy()
    df.columns = [str(c).strip() for c in df.columns]
    tx_col = None
    for c in df.columns:
        if "transaction" in c.lower() or "utr" in c.lower():
            tx_col = c
            break
    if tx_col is None:
        raise ValueError("CSV must include a Transaction ID / UTR column")

    df["_tx"] = df[tx_col].astype(str).str.strip()
    df = df[df["_tx"].ne("") & df["_tx"].ne("nan")]
    df = df.drop_duplicates(subset=["_tx"], keep="last")
    return df


def build_records(
    df: pd.DataFrame, valid_enrollment_ids: set[int], batch_map: dict[int, int | None]
) -> tuple[list[dict], int]:
    records: list[dict] = []
    skipped = 0

    for _, row in df.iterrows():
        tx = row["_tx"]
        enroll_raw = first_col(
            row,
            "enrollment_id",
            "enrollmentId",
            "Enrollment_Id",
            "Enrollment_No",
            "Enrollment No",
        )
        if enroll_raw is None:
            skipped += 1
            continue

        enroll_s = str(enroll_raw).strip()
        enrollment_id = None
        if enroll_s.lower() == "reco":
            enrollment_id = None
        elif re.fullmatch(r"\d+", enroll_s):
            eid = int(enroll_s)
            enrollment_id = eid if eid in valid_enrollment_ids else None
        else:
            enrollment_id = None

        amount = parse_float(
            first_col(row, "Transaction Amount (₹)", "Transaction Amount", "Amount", "amount")
        )
        discounted = parse_float(first_col(row, "Discounted Course Fee"))
        first_emi = parse_float(first_col(row, "1st EMI"))
        tenure = parse_int_optional(first_col(row, "tenure", "Tenure"))
        mode_val = first_col(row, "Mode", "mode")
        mode = str(mode_val).strip() if mode_val is not None else None
        date_val = parse_date(first_col(row, "Date", "date", "Payment Date"))

        records.append(
            {
                "transactionId": tx,
                "date": date_val,
                "enrollmentId": enrollment_id,
                "amount": amount,
                "mode": mode,
                "batchId": batch_map.get(enrollment_id) if enrollment_id else None,
                "discountedCourseFee": discounted if discounted else None,
                "firstEmi": first_emi if first_emi else None,
                "tenure": tenure,
            }
        )

    return records, skipped


def bulk_upsert_payment_table(
    engine: Engine, table: str, records: list[dict]
) -> int:
    if not records:
        return 0

    sql = text(
        f"""
        INSERT INTO `{table}` (
          transactionId, `date`, enrollmentId, amount, mode, batchId,
          discountedCourseFee, firstEmi, tenure, createdAt, updatedAt
        ) VALUES (
          :transactionId, :date, :enrollmentId, :amount, :mode, :batchId,
          :discountedCourseFee, :firstEmi, :tenure, NOW(), NOW()
        )
        ON DUPLICATE KEY UPDATE
          `date` = VALUES(`date`),
          enrollmentId = VALUES(enrollmentId),
          amount = VALUES(amount),
          mode = VALUES(mode),
          batchId = VALUES(batchId),
          discountedCourseFee = VALUES(discountedCourseFee),
          firstEmi = VALUES(firstEmi),
          tenure = VALUES(tenure),
          updatedAt = NOW()
        """
    )

    total = 0
    with engine.begin() as conn:
        for i in range(0, len(records), CHUNK_SIZE):
            chunk = records[i : i + CHUNK_SIZE]
            conn.execute(sql, chunk)
            total += len(chunk)
            print(f"  {table}: {min(i + CHUNK_SIZE, len(records))} / {len(records)}")
    return total


def bulk_upsert_consolidated(
    engine: Engine, records: list[dict], source_name: str
) -> int:
    if not records:
        return 0

    payload = [{**r, "sourceName": source_name} for r in records]
    sql = text(
        """
        INSERT INTO ConsolidatedPayment (
          transactionId, `date`, enrollmentId, amount, mode, batchId,
          discountedCourseFee, firstEmi, tenure, sourceName, createdAt, updatedAt
        ) VALUES (
          :transactionId, :date, :enrollmentId, :amount, :mode, :batchId,
          :discountedCourseFee, :firstEmi, :tenure, :sourceName, NOW(), NOW()
        )
        ON DUPLICATE KEY UPDATE
          `date` = VALUES(`date`),
          enrollmentId = VALUES(enrollmentId),
          amount = VALUES(amount),
          mode = VALUES(mode),
          batchId = VALUES(batchId),
          discountedCourseFee = VALUES(discountedCourseFee),
          firstEmi = VALUES(firstEmi),
          tenure = VALUES(tenure),
          sourceName = VALUES(sourceName),
          updatedAt = NOW()
        """
    )

    total = 0
    with engine.begin() as conn:
        for i in range(0, len(payload), CHUNK_SIZE):
            chunk = payload[i : i + CHUNK_SIZE]
            conn.execute(sql, chunk)
            total += len(chunk)
            print(f"  ConsolidatedPayment: {min(i + CHUNK_SIZE, len(payload))} / {len(payload)}")
    return total


def main() -> int:
    parser = argparse.ArgumentParser(description="Bulk import payment CSV to MySQL")
    parser.add_argument(
        "type",
        choices=sorted(PAYMENT_TABLES.keys()),
        help="Payment source type",
    )
    parser.add_argument("csv_path", help="Path to payment CSV file")
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Parse and validate only; do not write to database",
    )
    args = parser.parse_args()

    csv_path = Path(args.csv_path)
    if not csv_path.exists():
        print(f"File not found: {csv_path}", file=sys.stderr)
        return 1

    db_url = os.getenv("DATABASE_URL")
    if not db_url:
        print("DATABASE_URL not set in .env", file=sys.stderr)
        return 1

    table_name, source_label = PAYMENT_TABLES[args.type]
    engine = create_engine(
        prisma_url_to_sqlalchemy(db_url),
        pool_pre_ping=True,
        pool_size=5,
        max_overflow=10,
    )

    print(f"Reading {csv_path.name}...")
    df = pd.read_csv(csv_path, dtype=str, keep_default_na=True)
    df = normalize_csv(df)
    print(f"Unique transactions in CSV: {len(df)}")

    valid_ids, batch_map = load_reference_maps(engine)
    records, skipped = build_records(df, valid_ids, batch_map)
    print(f"Valid rows: {len(records)}, skipped: {skipped}")

    if args.dry_run:
        print("Dry run complete — no database writes.")
        return 0

    print(f"Uploading to {table_name} + ConsolidatedPayment...")
    bulk_upsert_payment_table(engine, table_name, records)
    bulk_upsert_consolidated(engine, records, source_label)

    with engine.connect() as conn:
        count = conn.execute(text(f"SELECT COUNT(*) FROM `{table_name}`")).scalar()
    print(f"Done. {table_name} total rows: {count}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/import-payment-bulk.mjs
/**
 * Fast bulk payment import via MySQL INSERT ... ON DUPLICATE KEY UPDATE.
 * ~100x faster than row-by-row Prisma upserts.
 *
 * Usage: node scripts/import-payment-bulk.mjs jodo "path/to/file.csv"
 */
import fs from 'fs';
import path from 'path';
import Papa from 'papaparse';
import { PrismaClient } from '../src/generated/prisma/index.js';

const prisma = new PrismaClient();

const SOURCE_MAP = {
  razorpay: { table: 'RazorpayPayment', delegate: 'razorpayPayment', label: 'Razorpay' },
  jodo: { table: 'JodoPayment', delegate: 'jodoPayment', label: 'Jodo' },
  early: { table: 'EarlyPayment', delegate: 'earlyPayment', label: 'Early' },
  offline: { table: 'OfflinePayment', delegate: 'offlinePayment', label: 'Offline' },
  bank: { table: 'BankPayment', delegate: 'bankPayment', label: 'Bank' },
  propelld: { table: 'PropelldPayment', delegate: 'propelldPayment', label: 'Propelld' },
  others: { table: 'OthersPayment', delegate: 'othersPayment', label: 'Others' },
};

const CHUNK = 2000;

function firstValue(row, keys) {
  for (const key of keys) {
    const v = row[key];
    if (v !== undefined && v !== null && String(v).trim() !== '') return v;
  }
  return null;
}

function parseDate(value) {
  if (!value) return null;
  const s = String(value).trim().split(/[ T]/)[0];
  const iso = s.match(/^(\d{4})-(\d{1,2})-(\d{1,2})$/);
  if (iso) {
    return `${iso[1]}-${iso[2].padStart(2, '0')}-${iso[3].padStart(2, '0')} 12:00:00`;
  }
  const dmy = s.match(/^(\d{1,2})-(\d{1,2})-(\d{4})$/);
  if (dmy) {
    return `${dmy[3]}-${dmy[2].padStart(2, '0')}-${dmy[1].padStart(2, '0')} 12:00:00`;
  }
  const d = new Date(value);
  if (!isNaN(d.getTime())) {
    return d.toISOString().slice(0, 19).replace('T', ' ');
  }
  return null;
}

function parseFloatVal(value) {
  if (value === null || value === undefined || value === '') return 0;
  return parseFloat(String(value).replace(/,/g, '')) || 0;
}

function parseIntOpt(value) {
  if (value === null || value === undefined || value === '') return null;
  const n = parseInt(String(value).trim(), 10);
  return Number.isNaN(n) ? null : n;
}

function sqlStr(v) {
  if (v === null || v === undefined) return 'NULL';
  return `'${String(v).replace(/\\/g, '\\\\').replace(/'/g, "''")}'`;
}

function sqlNum(v) {
  if (v === null || v === undefined) return 'NULL';
  return String(v);
}

async function bulkUpsert(table, rows) {
  if (!rows.length) return;
  const cols = `(transactionId, \`date\`, enrollmentId, amount, mode, batchId, discountedCourseFee, firstEmi, tenure, createdAt, updatedAt)`;
  const updates = `\`date\`=VALUES(\`date\`), enrollmentId=VALUES(enrollmentId), amount=VALUES(amount), mode=VALUES(mode), batchId=VALUES(batchId), discountedCourseFee=VALUES(discountedCourseFee), firstEmi=VALUES(firstEmi), tenure=VALUES(tenure), updatedAt=NOW()`;

  for (let i = 0; i < rows.length; i += CHUNK) {
    const chunk = rows.slice(i, i + CHUNK);
    const values = chunk
      .map(
        r =>
          `(${sqlStr(r.transactionId)}, ${r.date ? sqlStr(r.date) : 'NULL'}, ${sqlNum(r.enrollmentId)}, ${sqlNum(r.amount)}, ${r.mode ? sqlStr(r.mode) : 'NULL'}, ${sqlNum(r.batchId)}, ${sqlNum(r.discountedCourseFee)}, ${sqlNum(r.firstEmi)}, ${sqlNum(r.tenure)}, NOW(), NOW())`
      )
      .join(',\n');
    await prisma.$executeRawUnsafe(
      `INSERT INTO \`${table}\` ${cols} VALUES ${values} ON DUPLICATE KEY UPDATE ${updates}`
    );
    console.log(`  ${table}: ${Math.min(i + CHUNK, rows.length)} / ${rows.length}`);
  }
}

async function bulkUpsertConsolidated(rows, sourceName) {
  if (!rows.length) return;
  const cols = `(transactionId, \`date\`, enrollmentId, amount, mode, batchId, discountedCourseFee, firstEmi, tenure, sourceName, createdAt, updatedAt)`;
  const updates = `\`date\`=VALUES(\`date\`), enrollmentId=VALUES(enrollmentId), amount=VALUES(amount), mode=VALUES(mode), batchId=VALUES(batchId), discountedCourseFee=VALUES(discountedCourseFee), firstEmi=VALUES(firstEmi), tenure=VALUES(tenure), sourceName=VALUES(sourceName), updatedAt=NOW()`;

  for (let i = 0; i < rows.length; i += CHUNK) {
    const chunk = rows.slice(i, i + CHUNK);
    const values = chunk
      .map(
        r =>
          `(${sqlStr(r.transactionId)}, ${r.date ? sqlStr(r.date) : 'NULL'}, ${sqlNum(r.enrollmentId)}, ${sqlNum(r.amount)}, ${r.mode ? sqlStr(r.mode) : 'NULL'}, ${sqlNum(r.batchId)}, ${sqlNum(r.discountedCourseFee)}, ${sqlNum(r.firstEmi)}, ${sqlNum(r.tenure)}, ${sqlStr(sourceName)}, NOW(), NOW())`
      )
      .join(',\n');
    await prisma.$executeRawUnsafe(
      `INSERT INTO ConsolidatedPayment ${cols} VALUES ${values} ON DUPLICATE KEY UPDATE ${updates}`
    );
    console.log(`  ConsolidatedPayment: ${Math.min(i + CHUNK, rows.length)} / ${rows.length}`);
  }
}

async function main() {
  const type = process.argv[2];
  const filePath = process.argv[3];
  if (!type || !filePath || !SOURCE_MAP[type]) {
    console.error('Usage: node scripts/import-payment-bulk.mjs <jodo|razorpay|...> <csv-path>');
    process.exit(1);
  }

  const { table, delegate, label } = SOURCE_MAP[type];
  const text = fs.readFileSync(path.resolve(filePath), 'utf8');
  const parsed = Papa.parse(text.trim(), { header: true, skipEmptyLines: true });
  const byTx = new Map();

  for (const row of parsed.data) {
    const tx = firstValue(row, [
      'Settlement UTR / Transaction ID', 'TransactionId', 'transactionId',
      'Transaction ID', 'Transaction Id', 'transaction_id',
    ]);
    if (tx) byTx.set(String(tx), row);
  }

  const enrollRows = await prisma.enrollment.findMany({ select: { id: true, enrollment: true } });
  const validIds = new Set(enrollRows.map(r => r.id));
  const enrollTextToId = new Map(
    enrollRows.map(r => [String(r.enrollment).trim().toUpperCase(), r.id])
  );
  const formRows = await prisma.$queryRawUnsafe(`
    SELECT enrollment_no AS enrollmentId, batch AS batchId
    FROM AdmissionForm
    WHERE enrollment_no IS NOT NULL
  `);
  const batchMap = new Map(
    formRows.map((f) => [Number(f.enrollmentId), Number(f.batchId)])
  );

  function isRecoRow(row) {
    const idRaw = firstValue(row, [
      'enrollment_id', 'enrollmentId', 'Enrollment_Id', 'Enrollment ID', 'Enrollment_ID',
      'Enrollment_No', 'Enrollment No',
    ]);
    if (idRaw !== null && String(idRaw).trim().toLowerCase() === 'reco') return true;
    const textRaw = firstValue(row, ['Enrollment', 'EnrollmentNo', 'enrollmentNo', 'enrollment_no']);
    return textRaw !== null && String(textRaw).trim().toLowerCase() === 'reco';
  }

  function resolveEnrollmentId(row) {
    if (isRecoRow(row)) return null;
    const enrollRaw = firstValue(row, [
      'enrollment_id', 'enrollmentId', 'Enrollment_Id', 'Enrollment ID', 'Enrollment_ID',
      'Enrollment_No', 'Enrollment No',
    ]);
    if (enrollRaw !== null) {
      const s = String(enrollRaw).trim();
      if (s && s.toLowerCase() !== 'reco') {
        if (/^\d+$/.test(s)) {
          const id = parseInt(s, 10);
          return validIds.has(id) ? id : null;
        }
      }
    }
    const textRaw = firstValue(row, ['Enrollment', 'EnrollmentNo', 'enrollmentNo', 'enrollment_no']);
    if (textRaw !== null) {
      const s = String(textRaw).trim();
      if (s && s.toLowerCase() !== 'reco') {
        return enrollTextToId.get(s.toUpperCase()) ?? null;
      }
    }
    return null;
  }

  const records = [];
  let recoCount = 0;
  let skipped = 0;
  for (const row of byTx.values()) {
    const tx = firstValue(row, [
      'Settlement UTR / Transaction ID', 'TransactionId', 'transactionId',
      'Transaction ID', 'Transaction Id', 'transaction_id',
    ]);
    if (!tx) {
      skipped++;
      continue;
    }
    const enrollmentId = resolveEnrollmentId(row);
    if (enrollmentId === null && isRecoRow(row)) recoCount++;

    const discounted = parseFloatVal(row['Discounted Course Fee']);
    const firstEmi = parseFloatVal(row['1st EMI']);
    records.push({
      transactionId: String(tx),
      date: parseDate(firstValue(row, ['Date', 'date', 'Payment Date'])),
      enrollmentId,
      amount: parseFloatVal(firstValue(row, ['Transaction Amount (₹)', 'Transaction Amount', 'Amount', 'amount'])),
      mode: row.Mode || row.mode || null,
      batchId: enrollmentId ? (batchMap.get(enrollmentId) ?? null) : null,
      discountedCourseFee: discounted || null,
      firstEmi: firstEmi || null,
      tenure: parseIntOpt(firstValue(row, ['tenure', 'Tenure'])),
    });
  }

  console.log(`Bulk import: ${records.length} rows (${recoCount} reco → Reco tab, ${skipped} skipped) → ${table}`);
  await bulkUpsert(table, records);
  await bulkUpsertConsolidated(records, label);

  const count = await prisma[delegate].count();
  console.log('Done. Total in table:', count);
}

main()
  .catch(e => {
    console.error(e);
    process.exit(1);
  })
  .finally(() => prisma.$disconnect());
